engine/minskyTensorOps.cc

```diff
@@ -276,9 +280,59 @@ namespace minsky
   {
     std::shared_ptr<ITensor> arg1, arg2;
     void computeTensor() const override {//TODO
-      throw runtime_error("inner product not yet implemented");
+		
+	 if (arg1->rank()>2 || arg2->rank()>2)
+        throw runtime_error("inner product of rank>2 tensors not yet implemented");
+     
+     boost::numeric::ublas::matrix<double> m1 (arg1->hypercube().dims()[0],arg1->hypercube().dims()[1]);
+     if (arg1->rank()<2) boost::numeric::ublas::vector<double> m1 (arg1->hypercube().dims()[0]);
+     
+     for (size_t i=0; i<m1.size1(); ++i)
+         for (size_t j=0; j<m1.size2(); ++j)
+           if (arg1->rank()>1) m1(i,j)=arg1->atHCIndex(j+i*arg1->hypercube().dims()[arg1->rank()-1]);
+
+      if (arg1->rank()<2) {
+		  boost::numeric::ublas::vector<double> m1 (arg1->hypercube().dims()[0]);
+		  for (size_t i=0; i<m1.size(); ++i)
+             m1(i)=(*arg1)[i];   
+		 }
+              
+     boost::numeric::ublas::matrix<double> m2 (arg2->hypercube().dims()[0],arg2->hypercube().dims()[1]);
+     if (arg2->rank()<2) boost::numeric::ublas::vector<double> m2 (arg2->hypercube().dims()[0]);     
+     
+      for (size_t i=0; i<m2.size1(); ++i)
+          for (size_t j=0; j<m2.size2(); ++j)
+             if (arg2->rank()>1) m2(i,j)=arg2->atHCIndex(j+i*arg2->hypercube().dims()[0]);
+              
+      if (arg2->rank()<2) {
+		  boost::numeric::ublas::vector<double> m2 (arg2->hypercube().dims()[0]);
+		  for (size_t i=0; i<m2.size(); ++i)
+             m2(i)=(*arg2)[i];   
+		 }
+          
+     boost::numeric::ublas::matrix<double> dotProd=prec_prod(m1,m2);          
+	
+     size_t counter=0; 
+     for (size_t i=0; i<dotProd.size1(); ++i)
+         for (size_t j=0; j<dotProd.size2(); ++j, counter++)
+               cachedResult[counter]=dotProd(i,j);
+     
+     if (cachedResult.size()==0) 
+        for (size_t i=0; i<arg1->size(); i++) 
+           cachedResult[i]=nan("");
     }
     Timestamp timestamp() const override {return max(arg1->timestamp(), arg2->timestamp());}
+    void setArguments(const TensorPtr& a1, const TensorPtr& a2) override {
+      arg1=a1; arg2=a2;
+      if (arg1 && arg1->rank()!=0 && arg2 && arg2->rank()!=0) {
+        if (arg1->hypercube().dims()[arg1->rank()-1]!=arg2->hypercube().dims()[0])
+          throw std::runtime_error("arguments not conformal");
+        
+        // ensure size of final tensor is the same as the second argument, which can be rank 1 or 2  
+ 		cachedResult.hypercube(arg2->hypercube());
```

Comment:
The resultant hypercube has shape [i1,i2,...i{n-1},j2,j3,...j{m-1}], where ii are the dimensions of the first tensor and ji the dimensions of the second, n the rank of the first, and m the rank of the second. Not [j1,...jm] as you have here.

Reply:
I realized this yesterday when I was trying to get the matrix dot vector operation to work. I need to resize the hypercube to be a concatenation of the outer dimensions of the two tensors that comprise the inner product. I'll go and look at the civita class and the hypercube struct, in particular, to see how best to accomplish this.